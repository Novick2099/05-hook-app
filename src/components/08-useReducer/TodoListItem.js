import React from 'react';
import IconButton from "@material-ui/core/IconButton";
import DeleteIcon from "@mui/icons-material/Delete";
import {ListItem, ListItemButton, ListItemText} from "@mui/material";

const TodoListItem = ({key, todo, index, handleDelete, handleToggle}) => {

  const onClickToggle=()=>{
    handleToggle(todo.id)
  }

  return (
    <ListItem secondaryAction={
      <IconButton id={`delete_button_${index}`} edge="end" aria-label="delete" onClick={()=>{
        handleDelete(todo.id);
      }}>
        <DeleteIcon />
      </IconButton>}
    >
      <ListItemButton
        id={`TodoToggleButton${todo.id}}`}
        key={`TodoToggleButton${todo.id}}`}
        onClick={onClickToggle}>
        <ListItemText
          key={`TodoTextItem${todo.id}}`}
          primary={`${index+1}.${todo.desc}`}
          // todo preferir usar operadores ternarios a condition && value
          sx={todo.done ? {textDecoration: 'line-through'}:{}}
        />
      </ListItemButton>
    </ListItem>
  );
};

export default TodoListItem;
