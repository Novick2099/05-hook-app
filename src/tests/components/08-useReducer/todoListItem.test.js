import TodoListItem from "../../../components/08-useReducer/TodoListItem";
import {mount, shallow} from "enzyme";
import IconButton from "@material-ui/core/IconButton";
import {ListItem, ListItemButton, ListItemText} from "@mui/material";

describe('Test in TodoListItem',()=>{

  const newToDo = {
    id: 1,
    desc: 'Leer Java OCP 8',
    done: false,
  }

  const handleDelete = jest.fn();
  const handleToggle = jest.fn();

  const wrapper = mount(
    <TodoListItem
    key={1}
    todo={newToDo}
    index={0}
    handleDelete={handleDelete}
    handleToggle={handleToggle}
  />);


  it('Should match again snapshot',()=>{
    expect(wrapper).toMatchSnapshot();
    console.log(wrapper.debug());
  });

  it('Should call delete function', ()=>{
    const deleteButton = wrapper.find('[aria-label="delete"]').first();
    deleteButton.simulate('click');
    expect(handleDelete).toHaveBeenCalledWith(newToDo.id);
  })

  //todo Como simular click cuando onClick es un evento aÃ±adido listItem
  it('Should Call handleToggle',()=>{
    //todo 'button' funciona porque button es parte de html, pero con 'Button'
    // no funcionara porque esta es una etiqueta de react material
    // para buscar con la etiqueta de React usar find(EtiquetaReact)
    const listItemButton = wrapper.find(ListItemButton);
    const clickFunc=listItemButton.prop('onClick');
    clickFunc();
    expect(handleToggle).toHaveBeenCalled();
  });

  it('Should show the content in listItem',() => {
    const listItem = wrapper.find(ListItemText);
    expect(listItem.text()).toBe(`1.${newToDo.desc}`);
    expect(listItem.prop('sx')).toEqual({});
    expect(listItem.text()).toContain('1');
  });

});