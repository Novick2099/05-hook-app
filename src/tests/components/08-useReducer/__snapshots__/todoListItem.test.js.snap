// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in TodoListItem Should match again snapshot 1`] = `
Array [
  "",
  <li
    className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1rki8sg-MuiListItem-root"
    disabled={false}
  >
    Array [
      "",
      Array [
        "",
        <div
          className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          Array [
            "",
            <div
              className="MuiListItemText-root css-tlelie-MuiListItemText-root"
            >
              Array [
                "",
                <span
                  className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                >
                  NaN.undefined
                </span>,
              ]
            </div>,
          ]
          Array [
            "",
            <span
              className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />,
          ]
        </div>,
      ],
    ]
    Array [
      "",
      <div
        className="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
      >
        <button
          aria-label="delete"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
          disabled={false}
          id="delete_button_undefined"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            Array [
              "",
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>,
            ]
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>,
    ]
  </li>,
]
`;
